src/traversals/bfs.jl

```diff
@@ -170,3 +167,280 @@ function has_path(g::AbstractGraph{T}, u::Integer, v::Integer;
     end
     return false
 end
+
+using Base.Threads 
+
+"""
+    static_parallel_gdistances!(g, sources, vert_level; sort_alg=QuickSort)
+    static_parallel_gdistances!(g, source, vert_level; sort_alg=QuickSort)
+
+Parallel implementation of [`LightGraphs.gdistances!`](@ref) with static load balancing.
+
+### Performance
+Memory: `nthreads()*nv(g)*sizeof(eltype(g))`
+"""
+function static_parallel_gdistances!(
+    g::AbstractGraph{T}, 
+    sources::Vector{<:Integer},
+    vert_level::Vector{T};
+    sort_alg = Base.Sort.QuickSort
+    ) where T <:Integer
+ 
+    n = nv(g)
+    n_t = nthreads()
+
+    visited = zeros(Bool, n)
+    old_visited = zeros(Bool, n)
+#    visited = falses(n)
+#    old_visited = zeros(Bool, n)
+    fill!(vert_level, typemax(T))
+
+    # push! is not thread safe.
+    # Memory overhead could be reduced considerably if it were.
+    next_level_t = [Vector{T}(undef, n+1) for _ in 1:n_t]
+    vert_level = fill(typemax(T), n)
+    cur_level = Vector{T}(undef, length(sources))
+    cur_level .= sources
+    sizehint!(cur_level, n)
+
+    # End of queue is marked with 0
+    for t in 1:n_t
+        next_level_t[t][1] = zero(T)
+    end
+
+    for s in sources
+        visited[s] = old_visited[s] = true
+        vert_level[s] = zero(T)
+    end
+
+    n_level = one(T)
+    partitions = Vector{UnitRange{T}}(undef, n_t)
+
+    while !isempty(cur_level)
+        LightGraphs.unweighted_contiguous_partition!(partitions, length(cur_level))
+        
+        @threads for ind_set in partitions
+            t = threadid()
+            @inbounds next_level = next_level_t[t]
+            q_rear = zero(T) 
+            
+            # Explore vertices in each partition of cur_level.
+            @inbounds for ind in ind_set
+                v = cur_level[ind]
+                for i in outneighbors(g, v)
+                    if !visited[i]   # Data race, but first read always succeeds
+                        visited[i] = true
+                        q_rear+=one(T)
+                        next_level[q_rear] = i
+                    end
+                end
+            end
+
+            @inbounds next_level[q_rear+one(T)] = zero(T)  
+        end
+
+        # Update cur_level using next_level_t 
+        empty!(cur_level)
+        @inbounds for next_level in next_level_t
+
+            for v in next_level
+                (v == zero(T)) && break
+                if !old_visited[v] 
+                    push!(cur_level, v)
+                    old_visited[v] = visited[v] = true
+                    vert_level[v] = min(vert_level[v], n_level)
```

Comment:
If the `!old_visited[v]` is true for a vertex `v` how can `vert_level[v]` be set for it ?